datatypes/src/primitives/datetime.rs

```diff
@@ -0,0 +1,638 @@
+    }
+    pub fn hour(&self) -> u8 {
+        self.datetime.hour() as u8
+    }
+    pub fn minute(&self) -> u8 {
+        self.datetime.minute() as u8
+    }
+    pub fn second(&self) -> u8 {
+        self.datetime.second() as u8
+    }
+    pub fn millisecond(&self) -> u16 {
+        self.datetime.timestamp_subsec_millis() as u16
+    }
+
+    /// Creates a new `DateTime` from the original one by changing the year.
+    ///
+    /// Returns `None` if the year overflows the date.
+    pub fn with_year(&self, year: i32) -> Option<Self> {
+        Some(Self {
+            datetime: self.datetime.with_year(year)?,
+        })
+    }
+}
+
+impl FromStr for DateTime {
+    type Err = DateTimeError;
+
+    fn from_str(input: &str) -> Result<Self, Self::Err> {
+        let date_time = chrono::DateTime::<chrono::FixedOffset>::from_str(input).map_err(|e| {
+            DateTimeError::DateParse {
+                source: Box::new(e),
+            }
+        })?;
+
+        Ok(date_time.into())
+    }
+}
+
+impl Display for DateTime {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{}",
+            self.datetime
+                .to_rfc3339_opts(chrono::SecondsFormat::Millis, true)
+        )
+    }
+}
+
+impl From<DateTime> for TimeInstance {
+    fn from(datetime: DateTime) -> Self {
+        Self::from(&datetime)
+    }
+}
+
+impl From<&DateTime> for TimeInstance {
+    fn from(datetime: &DateTime) -> Self {
+        TimeInstance::from_millis_unchecked(datetime.datetime.timestamp_millis())
+    }
+}
+
+impl TryFrom<TimeInstance> for DateTime {
+    type Error = DateTimeError;
+
+    fn try_from(time_instance: TimeInstance) -> Result<Self, Self::Error> {
+        use chrono::TimeZone;
+
+        if time_instance < TimeInstance::MIN || time_instance > TimeInstance::MAX {
+            return Err(DateTimeError::OutOfBounds);
+        }
+
+        match chrono::Utc.timestamp_millis_opt(time_instance.inner()) {
+            chrono::LocalResult::Single(datetime) => Ok(Self { datetime }),
+            chrono::LocalResult::None | chrono::LocalResult::Ambiguous(_, _) => {
+                Err(DateTimeError::OutOfBounds)
+            }
+        }
+    }
+}
+
+impl From<chrono::DateTime<chrono::FixedOffset>> for DateTime {
+    fn from(datetime: chrono::DateTime<chrono::FixedOffset>) -> Self {
+        Self {
+            datetime: datetime.into(),
+        }
+    }
+}
+
+impl From<chrono::DateTime<chrono::Utc>> for DateTime {
+    fn from(datetime: chrono::DateTime<chrono::Utc>) -> Self {
+        Self { datetime }
+    }
+}
+
+impl From<DateTime> for chrono::DateTime<chrono::FixedOffset> {
+    fn from(datetime: DateTime) -> Self {
+        Self::from(&datetime)
+    }
+}
+
+impl From<&DateTime> for chrono::DateTime<chrono::FixedOffset> {
+    fn from(datetime: &DateTime) -> Self {
+        datetime.datetime.into()
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
```

Comment:
why is this not Deserialize? How is this constructed, aside from tests?

Reply:
There are custom impls